workspaces-tests/src/bin/simple_multisig.rs

```diff
@@ -0,0 +1,126 @@
+    },
+    rbac::Rbac,
+    slot::Slot,
+    Rbac,
+};
+use near_sdk::{
+    borsh::{self, BorshDeserialize, BorshSerialize},
+    env, near_bindgen, BorshStorageKey, PanicOnDefault,
+};
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKey {
+    SimpleMultisig,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+enum MyAction {
+    SayHello,
+    SayGoodbye,
+}
+
+impl approval::Action for MyAction {
+    type Output = &'static str;
+
+    fn execute(self) -> Self::Output {
+        match self {
+            Self::SayHello => "hello",
+            Self::SayGoodbye => "goodbye",
+        }
+    }
+}
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum Role {
+    Multisig,
+}
+
+#[derive(PanicOnDefault, BorshSerialize, BorshDeserialize, Rbac)]
+#[rbac(roles = "Role")]
+#[near_bindgen]
+pub struct Contract {}
+
+// This single function implementation completely implements simple multisig on
+// the contract
+impl ApprovalManager<MyAction, ApprovalState, Configuration<Self>> for Contract {
+    fn root() -> Slot<()> {
+        Slot::new(StorageKey::SimpleMultisig)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum ApproverError {
+    UnauthorizedAccount,
+}
+
+impl Display for ApproverError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Unauthorized account")
+    }
+}
+
+// We don't have to check env::predecessor_account_id or anything like that
+// SimpleMultisig handles it all for us
+impl AccountAuthorizer for Contract {
+    fn is_account_authorized(account_id: &near_sdk::AccountId) -> bool {
+        Contract::has_role(account_id, &Role::Multisig)
+    }
+}
+
+#[near_bindgen]
+impl Contract {
+    const APPROVAL_THRESHOLD: u8 = 2;
+    const VALIDITY_PERIOD: u64 = 1000000 * 1000 * 60 * 60 * 24 * 7;
+
+    #[init]
+    pub fn new() -> Self {
+        <Self as ApprovalManager<_, _, _>>::init(Configuration::new(
+            Self::APPROVAL_THRESHOLD,
+            Self::VALIDITY_PERIOD,
+        ));
+
+        Self {}
+    }
+
+    pub fn obtain_multisig_permission(&mut self) {
+        self.add_role(&env::predecessor_account_id(), &Role::Multisig);
+    }
+
+    pub fn request(&mut self, action: String) -> u32 {
+        let action = match &action[..] {
+            "hello" => MyAction::SayHello,
+            "goodbye" => MyAction::SayGoodbye,
+            _ => env::panic_str("action must be \"hello\" or \"goodbye\""),
+        };
+
+        let request_id = self.create_request(action, ApprovalState::new()).unwrap();
+
+        near_sdk::log!(format!("Request ID: {request_id}"));
+
+        request_id
+    }
+
+    pub fn approve(&mut self, request_id: u32) {
+        self.approve_request(request_id).unwrap();
+    }
+
+    pub fn is_approved(&self, request_id: u32) -> bool {
+        <Contract as ApprovalManager<_, _, _>>::is_approved(request_id)
+    }
+
+    pub fn execute(&mut self, request_id: u32) -> String {
+        self.execute_request(request_id).unwrap().to_string()
+    }
```

Comment:
@encody I was kind of surprised that this section of thin wrappers was necessary.

Is there a for us to offer NCT in a way where the users don't even need to code these wrapper functions?

Also, I'd like a better example than `obtain_multisig_permission` (at least in the docs, if not here in the tests) because this feels misleading.  Should we show an example where only the contract owner can add accounts to a role?

Lastly, Denys mentioned that he often sees elsewhere the concept of "request and approve" and "approve and execute" combo functions. Should we offer those?

Reply:
> Is there a for us to offer NCT in a way where the users don't even need to code these wrapper functions?

If the module implements public functions it requires an NEP. As written, the module would not currently require an NEP for its pubic interface, since it does not have one. I don't think we've agreed to create an NEP for multisig yet, so I have refrained from writing code under that assumption.

> Also, I'd like a better example than `obtain_multisig_permission` (at least in the docs, if not here in the tests) because this feels misleading. Should we show an example where only the contract owner can add accounts to a role?

These tests are primarily intended to test modules, not demonstrate fully-featured blockchain applications. Such a set of example contracts could be written, however.

> Lastly, Denys mentioned that he often sees elsewhere the concept of "request and approve" and "approve and execute" combo functions. Should we offer those?

No. The functions provided can be sequenced to perform the exact same operation.